MAVProxy/modules/mavproxy_steamvr.py

```diff
@@ -0,0 +1,303 @@
+                frame_count = 0
+                gps_frame_count = 0
+                vision_frame_count = 0
+
+
+            # check if the pose is valid
+            if pose.eTrackingResult != vr.TrackingResult_Running_OK or not pose.bPoseIsValid or not pose.bDeviceIsConnected:
+                self.health = False
+                continue
+            self.health = True
+
+            # read in position and attitude vector
+            pose_mat = pose.mDeviceToAbsoluteTracking
+
+            # convert to NED
+            pos_ned = Vector3(-pose_mat[2][3], pose_mat[0][3], -pose_mat[1][3])
+
+            # extract the 3x3 rotation matrix
+            rot_mat = np.array([ [pose_mat[0][0], pose_mat[0][1], pose_mat[0][2]],[pose_mat[1][0], pose_mat[1][1], pose_mat[1][2]],[pose_mat[2][0], pose_mat[2][1], pose_mat[2][2]] ])
+
+            # rotate about x
+            x_mat = np.array([ [1,0,0], [0,0,1], [0,-1,0] ])
+
+            mat = rot_mat.dot(x_mat)
+
+            # get the quaternion
+            q = Quaternion(mat)
+
+            # Convert quaternion to roll, pitch and yaw
+            # For some reason the built in method does not return the expected result
+            # it is highly likely that the DCM is rotated in some intresting way
+            # hard to tell because the built in function goes via DCM and the working example is direct
+            # the built in DCM to quaternion function is identical
+            [pitch,roll,yaw] = self.tracker_q_to_euler(q[0],q[1],q[2],q[3])
+            pitch *= -1
+            yaw *= -1
+            yaw += math.radians(self.vr_settings.yaw_offset + 180)
+            yaw = math.radians(mavextra.wrap_360(math.degrees(yaw)))
+
+            # we also have acesees to the rotation rates in rads/s
+            # I have not worked out the conversion as it is not currently used
+            #rot_rate = pose.vAngularVelocity
+
+            self.pos = pos_ned
+            self.att = [math.degrees(roll), math.degrees(pitch), math.degrees(yaw)]
+
+            # read in velocity's
+            vel_in = pose.vVelocity
+            vel_ned = Vector3(-vel_in[2], vel_in[0], -vel_in[1])
+            self.vel = vel_ned
+
+            # match requested send rate
+            if self.vr_settings.vision_rate > 0:
+                dt = now - last_msg_time
+                if dt < 1.0 / self.vr_settings.vision_rate:
+                    continue
+            last_msg_time = now
+
+            mav = self.master
+            time_us = int(now * 1.0e6)
+            if now - last_origin_send > 1 and self.vr_settings.vision_rate > 0:
+                # send a heartbeat msg
+                mav.mav.heartbeat_send(mavutil.mavlink.MAV_TYPE_GCS, mavutil.mavlink.MAV_AUTOPILOT_GENERIC, 0, 0, 0)
+
+                # send origin at 1Hz
+                mav.mav.set_gps_global_origin_send(self.target_system,
+                                                   int(self.vr_settings.origin_lat*1.0e7),
```

Comment:
I thought this only had to be sent once?

Reply:
It only has to be received once, I guess this is easier than trying to tell. This is completely left over from the Vicon stuff.